savefile/savefile.go

```diff
@@ -0,0 +1,129 @@
+package savefile
+
+import (
+	"bytes"
+	"compress/zlib"
+	"encoding/binary"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	"github.com/sector-f/jhmod/nvc"
+)
+
+const (
+	// Magic header for Jupiter Hell save save files.
+	magic = "\xde\xc0\xad\xde"
+)
+
+type savedata struct {
+	// Player name
+	PlayerName string
+	// Game mode.  This can be "jh", and various others.
+	GameMode string
+	// The current level's name.
+	CurrentLevel string
+	// The seed used to generate the game.
+	Seed uint32
+}
+
+// Something was reading for data but got an EOF.
+var UnexpectedEndOfFileErr error = errors.New("unexpected eof")
+
+// General parse error.  TODO make more descriptive errors.
+var SaveParseErr error = errors.New("general parse error")
+
+// Read in a string from a buffer of size suffixed with a length byte.
+func readString(r io.Reader, size int) (string, error) {
+	buf := make([]byte, size)
+	nRead, err := r.Read(buf[:])
+	if err != nil {
+		return "", err
+	}
+	if nRead != size {
+		return "", UnexpectedEndOfFileErr
+	}
+	recordedSize := buf[size-1]
+	if int(recordedSize) > size {
+		return "", SaveParseErr
+	}
+	stringBytes := buf[:recordedSize]
+
+	for _, b := range stringBytes {
+		if b == 0 {
+			return "", SaveParseErr
+		}
+	}
```

Comment:
Could you add a comment explaining why you're specifically checking `if b == 0`?

I'm _assuming_ it's because valid text (almost certainly) won't contain NUL characters, but I'm not 100% sure (and there are other ASCII characters that almost certainly won't appear in "valid text", e.g. file/group separators, start of text, etc). Maybe I'm missing something!

Reply:
Yep valid ascii.  I'll update it to do that.